### PR TITLE
Use data attribute instead of id for cell tagging

### PR DESCRIPTION
**Pull request by mmkal**
Hi - opening this up for the sake of discussion initially - and hopefully eventually merging. I didn't spot this in #42 but I think it's worth addressing one way or another (even if it's not exactly this solution).

This change stops using [id](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) on cell elements. Instead it now uses a custom [data-* attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/data-*).

The main reason for this being that this isn't what an `id` is intended for - it's supposed to globally, uniquely, tag an HTML element. This could cause issues on pages that want to render multiple crosswords (something we are planning on - we want to have a page which renders several scrollable quick-fire crosswords stacked on top of each other so players can challenge themselves to see how fast they can scroll through all of them). It's a big no-no to have two elements with duplicate `id`s (although browsers will render them). Even if crosswords-js is careful to avoid the pitfalls of having non-unique `id`s, applications and testing libraries like [playwright](https://playwright.dev) could still fall foul of them. I remember a discussion in one of the PRs about `id`s - I thought the upshot was they make sense in the sample app, but not as part of the library.

A second reason is that having `id`s that use special characters (`(`, `,` and `)`) are at odds with almost all `id`s I've seen in my past, and all the styleguides/naming conventions I could find from googling. So I think this change is better in terms of the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

Google's result for "html data attribute vs id" summarises this quite well:

image

The attribute looks like data-xy="3,4". I also went from (3,4) to 3,4 because:

- that matches what squares.io does. Like it or not, speaking with setters and solvers alike, squares.io is used by many for general experimentation. I don't think going for 100% parity is necessary/ a good use of effort, but given I'm proposing a change already, may as well nudge in the direction of friendliness/compatibility.
- It makes the HTML and code relying on it more readable. I have some playwright tests on my web app that do things like `await page.click('[data-xy="3,4"]')` - there is less confusing punctuation there than with `await page.click('[data-xy="(3,4)"]')`

Squares.io example:

image

crosswords-js example with this change:

image

Some other changes along the way:

- added jsdoc types to a few methods related to cell-tagging

Implementation note:

- I used the native [`dataset` property](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset) to get and set the `data-xy` value, it seems to be the recommended way over `cellElement.setAttribute('data-xy', modelCell)` - but they're equivalent.
- I updated the `input` data-xy value too, although I don't think it's really necessary. It's easy to access the input element via a selector like `document.querySelector('[data-xy="3,4"] input')`

Happy to discuss alternatives - but I do think this is an important change for this library to get to a stable version (v1) - otherwise people will start relying on `id` and it'll be very hard to change.

### CODE REVIEW

**Comment by mmkal** [comment]
Oh @pvspain I meant to ask, do you remember what the problem was that led to `'id' is not used, but assignment silences chromium-dev-tools issue` - and adding the `input-${modelCell}` id to the input elements? I should make sure this doesn't regress that problem (and see if we can remove the tag there without regressing it).

**Comment by pvspain** [comment]
> Hi - opening this up for the sake of discussion initially - and hopefully eventually merging. I didn't spot this in #42 but I think it's worth addressing one way or another (even if it's not exactly this solution).
> 
> This change stops using [id](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) on cell elements. Instead it now uses a custom [data-* attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/data-*).
> 
> The main reason for this being that this isn't what an `id` is intended for - it's supposed to globally, uniquely, tag an HTML element.  
...
Even if crosswords-js is careful to avoid the pitfalls of having non-unique `id`s, applications and testing libraries like [playwright](https://playwright.dev) could still fall foul of them. I remember a discussion in one of the PRs about `id`s - I thought the upshot was they make sense in the sample app, but not as part of the library.
> 

In the earlier discussion I recall floating the idea of using a gridview-instance-unique prefix for its `id`s to address the multi-gridview webpage. Something along the lines of a namespace-_ish_ tag (e.g. `cwjs`) with a singleton generator sequence value like [this](https://stackoverflow.com/a/1997811/1090683) tacked onto it. That's just a strawman or talking point.

> A second reason is that having `id`s that use special characters (`(`, `,` and `)`) are at odds with almost all `id`s I've seen in my past, and all the styleguides/naming conventions I could find from googling. So I think this change is better in terms of the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).


Ids were introduced into the code to optimise the performance of the [`CellMap`](https://github.com/dwmkerr/crosswords-js/blob/main/src/cell-map.mjs) class. 

That class is used by the _grid element helpers_ in the _controller_, which are used heavily throughout the _controller_ code. Originally, the (_modelCell_,_cellElement_) mappings were stored in an array, and retrieving a mapping in either direction was achieved by a linear scan of that array.

I changed the implementation so the _cellElements_ became properties on a internal _modelCells_ object within a _CellMap_, and we could leverage the JS object-property lookup, which is presumably a hash-like arrangement and has close to constant-time efficiency. This required the _cellElement_ (the key - _modelCells.{property-name}_) to have an (unique) id value. I reasoned the best id for a _cellElement_ in this context was the id for it's mapped _modelCell_, so mapping a _cellElement_ to it's _modelCell_ simplified to a property reference.

I had previously added `.toString()` methods to _clue_ and _modelCell_, which work something like Python's [dynamic attributes](https://bjpcjp.github.io/pdfs/python/fluent-python-ch19.pdf). This handy feature _implicitly_ casts an Object to its string value when the context requires a string, such as in parameterised strings. I use this _a lot_ in console messages. For the console message context, I chose the _cell_'s grid coordinates, `(x,y)`.

Following the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle), I thought these methods would be a quick and cheap solution to the **id** requirement in the new `CellMap`.

> The attribute looks like data-xy="3,4". I also went from (3,4) to 3,4 because:
> 
> * that matches what squares.io does. Like it or not, speaking with setters and solvers alike, squares.io is used by many for general experimentation. I don't think going for 100% parity is necessary/ a good use of effort, but given I'm proposing a change already, may as well nudge in the direction of friendliness/compatibility.
> * It makes the HTML and code relying on it more readable. I have some playwright tests on my web app that do things like `await page.click('[data-xy="3,4"]')` - there is less confusing punctuation there than with `await page.click('[data-xy="(3,4)"]')`
> 

The _HTML-id_ context is quite different to the _CellMap_ and _console-message_ contexts, as you have elaborated. Your _HTML-id_ context falls squarely under the scope of the HTML `dataset` property as you've described. I feel your notation is too verbose when you're scanning console messages spewing into a narrow column!

**I don't think the different contexts need to use the same object property (`id`)**.  As we have different contexts, perhaps neither context should use the unadorned `id` - to prevent ambiguity. On second thought, I guess you have no choice but to use the `id` property on a HTML element class.  haven't looked into your implementation yet, so I'll shut up now and do that, as I may be wandering off on a tangent! 

> Happy to discuss alternatives - but I do think this is an important change for this library to get to a stable version (v1) - otherwise people will start relying on `id` and it'll be very hard to change.

Agreed.

**Comment by pvspain** [comment]
> Oh @pvspain I meant to ask, do you remember what the problem was that led to `'id' is not used, but assignment silences chromium-dev-tools issue` - and adding the `input-${modelCell}` id to the input elements? I should make sure this doesn't regress that problem (and see if we can remove the tag there without regressing it).

The first _Issues_ tab entry in the screencap of the Chromium Dev Tools...
Just trying to reduce the signal-to-noise ratio in reported problems!
Screenshot 2023-10-26 22 27 05

**Comment by mmkal** [comment]
Thank you, useful context. It might be helpful to go over the use cases/concerns for this tag:

### Performance in this library

I would be very surprised if this PR regressed performance within crosswords-js. The only changes are `cellElement.id` -> `cellElement.dataset.xy`. Both should be O(1) lookups. We're just looking up in a hashmap in `CellMap` as you say, so there should be no difference here.

### Performance in userland (via `document.querySelector`)

If library users are relying on anything like `document.getElementById('(3,4)')`, they would have to switch to `document.querySelector('[data-xy="3,4"]')` which might be a little slower. But, from what I read online, the difference should be infinitesimal and the latter should still be executed at a rate in the order of millions of queries per second, so I don't think this will ever going to be a bottleneck, unless someone builds a crossword with literally thousands of rows and columns (an interesting use case actually!). Also there is nothing stopping an app user adding their own ids to the cell elements in a subclass of this constructor (I am doing something like this to get the `data-xy` attribute already). And I think that's appropriate - IMO `id`s should be set in apps, not libraries. So I don't think this should be a concern.

### Debug logging

Yeah, let's make sure this doesn't get overly verbose. This is what they look like with this change (`currentCell: 1,0` etc.), I think it's pretty compact + readable?

image

>too verbose

Does the above cover your concern re verbosity, or did I misunderstand which logs you meant?

### Chrome dev tools issues

Ah - thank you for the pointer! Just seen `A form field element has neither an id nor a name attribute`. I just pushed a commit 6cb7c9b8613885cd724ceffad857452d27c3c925 which addresses by setting the `name` attribute rather than `id`. I think that's more appropriate since there's no expectation for that to be unique. I confirmed that all the issues go away when running `npm run dev`.

File: src/crossword-gridview.mjs [hunk 1]
```diff
@@ -78,7 +78,7 @@ function newCellElement(document, modelCell) {
   //  Light cells also need an input.
   const inputElement = document.createElement('input');
   // 'id' is not used, but assignment silences chromium-dev-tools issue.
```
Comment (pvspain) [inline src/crossword-gridview.mjs]: Please maintain comment above change: 'id' -> 'name'

**Review by pvspain — APPROVED** [review]
Hi @mmkal , This looks good - thank you!
There's just one little tweak below where you need to change the comment above to reflect the change.

**Comment by mmkal** [comment]
@dwmkerr bumping this - could we merge, if no objections?

**Comment by dwmkerr** [comment]
Looks good to me, merging now, thanks!

**Pull request merged**